onnx/helper.py

```diff
@@ -282,22 +282,23 @@ def split_complex_to_pairs(ca: Sequence[np.complex64]) -> Sequence[int]:
             for i in range(len(ca) * 2)]
 
 
-# convert a f32 to bf16 (as int)
-def float32_to_bfloat16(fval: float) -> int:
+# convert a float32 value to a bfloat16 (as int)
+# By default, this conversion rounds-to-nearest-even and supports NaN
+# Setting `truncate` to True enables a simpler conversion. In this mode the
+# conversion is performed by simply dropping the 2 least significant bytes of
+# the significand. In this mode an error of up to 1 bit may be introduced and
+# preservation of NaN values is not be guaranteed.
+def float32_to_bfloat16(fval: float, truncate: bool = False) -> int:
     ival = int.from_bytes(struct.pack('<f', fval), 'little')
+    if truncate:
+        return ival >> 16
+    # NaN requires at least 1 significand bit set
     if isnan(fval):
-        # NaN requires at least 1 significand bit set
-        ival16 = 0x7FC0  # sign=0, exp=all-ones, sig=0b1000000
-    else:
-        # drop bottom 16-bits
-        # round remaining bits using round-to-nearest-even
-        round = ((ival >> 16) & 1) + 0x7fff
-        ival16 = (ival + round) >> 16
-    # swap byte order for big-endian
-    if sys.byteorder == 'big':
-        bytes = struct.pack('<h', ival16)
-        ival16 = int.from_bytes(bytes, 'big')
-    return ival16
+        return 0x7FC0  # sign=0, exp=all-ones, sig=0b1000000
+    # drop bottom 16-bits
+    # round remaining bits using round-to-nearest-even
+    round = ((ival >> 16) & 1) + 0x7fff
+    return (ival + round) >> 16
```

Comment:
I assume that the endian adjustment was removed intentionally (and that it is not required)?

Reply:
I suspect no special-casing is required for endian because you are using "integer arithmetic operators" (and implementing them correctly is the compiler's responsibility, regardless of endian-ness). I mean, the operators used have a fixed semantics, regardless of the endian-ness.

Reply:
> Out of curiosity: I don't follow the `(ival >> 16) & 1` bit ... we end up adding either 0x7fff or 0x8000 (depending on the 16th bit) ... which seems slightly odd.

this is the "even" part of the "round-to-nearest-even" rounding. In this rounding mode if the fractional part being dropped is half-way between the two potential target values we round toward the nearest even value.

Specifically, we partition ival into two 16-bit parts [A][B]. If B is less than 0x8000 we round A down. if B is greater than 0x8000 we round A up. if B is exactly equal to 0x8000 we round A up or down based on if A is odd or even.


if A is odd `(ival >> 16) & 1` is 1, so `round` will be `0x8000`
if A is even `(ival >> 16) & 1` is 0, so `round` will be `0x7FFFF`

This lines up with what you noticed above, that round is always one of these two values.

Now the key is that we add this to the full 32-bit number.

If B is < 0x8000 and we add either of these values, the result will not overflow into A, so the addition is a no-op. When we drop B, we have effectively "rounded down"

if B is > 08x000 and we add either of these values, the result will *always* oveerflow a single bit into A. When we drop B, we have effectively "rounded up"

The magic is if B is exactly equal to 0x8000. 
- if we add 0x7FFF to this value, we do *not* overflow, the effect is round-down.
- if we add 0x8000 to this value, we do overflow, the effect is round-up.

So, if A is even, we round down (to the nearest even) and if A is odd, we round up (to the nearest even).

A useful analogy, that might be more familiar is rounding values in C when converting from FP to INT. In C this conversion "truncates" by dropping the fractional portion effectively implementing "round-toward-zero". However, if you add 0.5 to the value before converting, you get a "round-to-nearest" rounding. The difference between that technique and what i've implemented here is that we don't want "round-to-nearest" we want "round-to-nearest-even" so we need an extra step to force rounding down when the value is already even.


I hope this explanation makes sense!

Reply:
> I suspect no special-casing is required for endian because you are using "integer arithmetic operators" (and implementing them correctly is the compiler's responsibility, regardless of endian-ness). I mean, the operators used have a fixed semantics, regardless of the endian-ness.

I suspect also that internally python is strong its values in little-endian format regardless of the underlying machine, and that was confusing me. In otherwards i was overcomplicating things by conflating what i would need to do in C vs. what the semantics were of what i wanted. High-level languages FTW win here i suppose.

Reply:
Thanks for the detailed explanation about the rounding.